### PR TITLE
[WIP] feat: group server options

### LINKED ISSUE #3074: Grouping server options
## 🚀 Feature Proposal

Update the server configuration input to provide a clearer categorization of the options

## Motivation

Increase the focus of some properties; maybe a doc revision could be enough?

## Example

```js
fastify({
  timeouts: {
    connectionTimeout
    keepAliveTimeout
    pluginTimeout
    http2SessionTimeout
  },
  routing: {
    rewriteUrl
    constraints
    exposeHeadRoutes
    caseSensitive
    ignoreTrailingSlash
  },
  security: {
    onProtoPoisoning
    onConstructorPoisoning
    trustProxy
  },
  errorManagement: {
     clientErrorHandler
     frameworkErrors
     return503OnClosing
     schemaErrorFormatter
  }
  etc...
})
```

### PR DESCRIPTION
**Pull request by meehawk**
#### Description

This PR closes #3074

TODOS:

- [x] Add a helper function that normalizes options to the new grouped format
- [ ] Display deprecation warnings for old format of the options
- [x] Fix setting of default values post normalization
- [x] Fix schema for server options in build-validation.js 
- [x] Fix corresponding type definitions if any
- [x] Update tests to use the new grouped version for options 
- [ ] Update docs

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: fastify.js [hunk 1]
```diff
@@ -629,6 +644,50 @@ function fastify (options) {
   }
 }
 
+function normalizeOptions (options) {
+  const deprecatedOptionsMap = new Map([
```
Comment (Eomm) [inline fastify.js:648]: I would add these settings to `log`:

```
  logger
  requestIdLogLabel
  disableRequestLogging
  requestIdHeader
  genReqId
```

File: fastify.js [hunk 2]
```diff
@@ -121,21 +123,34 @@ function fastify (options) {
...
+  options.errorManagement.clientErrorHandler = options.errorManagement.clientErrorHandler || defaultClientErrorHandler
+  options.routing.exposeHeadRoutes = exposeHeadRoutes
+
+  const {
+    errorManagement,
+    routing,
+    security,
+    timeouts,
+    ...rest
+  } = options
```
Comment (Eomm) [inline fastify.js:144]: I would avoid this desctructure only to use it later

File: fastify.js [hunk 3]
```diff
@@ -629,6 +644,50 @@ function fastify (options) {
...
+
+  for (const key in options) {
+    if (deprecatedOptionsMap.has(key)) {
+      // !TODO: Print deprecation warning
+      const groupKey = deprecatedOptionsMap.get(key)
+
+      normalizedOptions[groupKey] = {
+        ...normalizedOptions[groupKey],
+        [key]: options[key]
+      }
```
Comment (Eomm) [inline fastify.js:677]:
```suggestion
      normalizedOptions[groupKey][key] = options[key]
```

File: fastify.js [hunk 4]
```diff
@@ -629,6 +644,50 @@ function fastify (options) {
...
+    ['onProtoPoisoning', 'security'],
+    ['onConstructorPoisoning', 'security'],
+    ['trustProxy', 'security'],
+    ['connectionTimeout', 'timeouts'],
+    ['keepAliveTimeout', 'timeouts'],
+    ['pluginTimeout', 'timeouts'],
+    ['http2SessionTimeout', 'timeouts']
+  ])
+
+  const normalizedOptions = {}
```
Comment (Eomm) [inline fastify.js:667]:
```suggestion
  const normalizedOptions = {
    errorManagement: {},
    routing: {},
    security: {},
    timeouts: {}
  }
```

File: fastify.js [hunk 5]
```diff
@@ -629,6 +644,50 @@ function fastify (options) {
...
+      delete options[key]
+    }
+  }
+
+  return Object.assign({
+    errorManagement: {},
+    routing: {},
+    security: {},
+    timeouts: {}
+  }, normalizedOptions, options)
```
Comment (Eomm) [inline fastify.js:688]:
```suggestion
  return Object.assign(normalizedOptions, options)
```

**Review by Eomm — COMMENTED** [review]
I think the docs need to be updated aswell

File: fastify.js [hunk 6]
```diff
@@ -121,21 +123,34 @@ function fastify (options) {
...
+    timeouts,
+    ...rest
+  } = options
+  const initialConfig = getSecuredInitialConfig({
+    ...errorManagement,
+    ...routing,
+    ...security,
+    ...timeouts,
+    ...rest
+  })
```
Comment (mcollina) [inline fastify.js:151]: Please don't do this. We should use the new format for this not the old one.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: fastify.js (hunk 6 above)
Comment (meehawk) [reply fastify.js:151]: @mcollina, this would also imply a corresponding change in schema for validation in ajv right?

File: fastify.js (hunk 2 above)
Comment (meehawk) [reply fastify.js:144]: Ahh I think the complete logic here would need to be scrapped as per @mcollina 's comment below. I was doing this to kind of use the schema validation in its old format as is

File: fastify.js (hunk 1 above)
Comment (meehawk) [reply fastify.js:648]: Thanks for pointing this out. I will add another field for these

File: fastify.js (hunk 6 above)
Comment (mcollina) [reply fastify.js:151]: Yes exactly.

Essentially:

1. transform old options to new ones, adding warnings for each one of them.
2. validate
3. use new options everywhere.

Comment (meehawk) [reply fastify.js:151]: Gotcha, right on it

**Comment by meehawk** [comment]
@mcollina sorry for the inactivity, I have some college stuff. I will start working on this by July 2nd

**Comment by meehawk** [comment]
@mcollina Sorry for slowing down on the PR. I've had a surgery recently and have been recuperating. I've lost a bit of context so I'll prolly have to spend some time going over the PR and making the changes. Will do that over the weekend

**Comment by zekth** [comment]
> @mcollina Sorry for slowing down on the PR. I've had a surgery recently and have been recuperating. I've lost a bit of context so I'll prolly have to spend some time going over the PR and making the changes. Will do that over the weekend

Take a rest, there is no rush to land to quickly. Also if you need help feel free to ask for it.

**Pull request closed without merging**